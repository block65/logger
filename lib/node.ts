import type { RequestHandler } from 'express';
import { serializeError } from 'serialize-error';
import * as pino from 'pino';
import * as pinoHttp from 'pino-http';
import { createNamespace, Namespace } from 'cls-hooked';
import { hostname } from 'os';
import { randomBytes } from 'crypto';
import { IncomingMessage, ServerResponse } from 'http';
import { sep } from 'path';

type Falsy = false | undefined | null;
type MixinFnWithData = (
  data: ReturnType<pino.MixinFn>,
) => ReturnType<pino.MixinFn>;

export type Logger = {
  cls: Namespace;
} & pino.Logger;

interface ClsContext {
  _contextId: string;
  _context?: Record<string, unknown>;
}

type NamespaceContext = {
  id?: unknown;
  _ns_name?: unknown;
} & Partial<ClsContext>;

type ComputePlatform = 'gcp-cloudrun' | 'aws-lambda' | 'aws';

export interface CreateLoggerOptions extends Omit<pino.LoggerOptions, 'mixin'> {
  traceCaller?: boolean;
  platform?: ComputePlatform;
  mixins?: (MixinFnWithData | pino.MixinFn | Falsy)[];
}

const suggestedHostname = hostname();

const defaultLoggerOptions: pino.LoggerOptions = {
  level: 'info',
  serializers: {
    err: serializeError,
  },
  base: {
    pid: process.pid,
    // in the case we get localhost (gcp cloud run), use random bytes instead
    // this helps us debug if we need to isolate calls running in a specific
    // container or instance (one which stays alive between invocations)
    hostname:
      suggestedHostname !== 'localhost'
        ? suggestedHostname
        : `zz${randomBytes(5).toString('hex')}`, // zz indicates synthetic hostname
  },
};

const gcpLevelMap: Record<string, string> = {
  trace: 'DEFAULT',
  debug: 'DEBUG',
  info: 'INFO',
  warn: 'WARNING',
  error: 'ERROR',
  fatal: 'CRITICAL',
};

function getPlatformLoggerOptions(
  platform?: ComputePlatform,
): pino.LoggerOptions {
  switch (platform) {
    case 'gcp-cloudrun':
      return {
        level: 'info',
        // customLevels: {
        //   DEBUG: 100,
        //   INFO: 200,
        //   NOTICE: 300,
        //   WARNING: 400,
        //   ERROR: 500,
        //   CRITICAL: 600,
        //   ALERT: 700,
        //   EMERGENCY: 800,
        // },
        messageKey: 'message',
        formatters: {
          level(levelLabel /* , levelNumber */) {
            return { severity: gcpLevelMap[levelLabel] };
          },
        },
        timestamp: pino.stdTimeFunctions.isoTime,
      };
    case 'aws':
    case 'aws-lambda':
    default:
      return {};
  }
}

let counter = 0;

const isDevelopment = process.env.NODE_ENV === 'development';

function callerMixin(): { caller: string | undefined } {
  return {
    caller: Error()
      .stack?.split('\n')
      .slice(2)
      .filter(
        (s) =>
          // !s.includes(`api${sep}src${sep}app`) &&
          // !s.includes(
          //   `node_modules${sep}express${sep}lib${sep}router`,
          // ) &&
          !s.includes(`node_modules${sep}pino`) &&
          !s.includes(`block65${sep}logger`),
      )?.[0]
      ?.substr(7),
  };
}

function composeMixins(
  mixins: (MixinFnWithData | pino.MixinFn | Falsy)[],
): pino.MixinFn {
  return () =>
    mixins.reduce((accum, mixin) => {
      if (!mixin) {
        return accum;
      }
      return {
        ...accum,
        ...mixin(accum),
      };
    }, {});
}

function createContextMixin(cls: Namespace): pino.MixinFn {
  return (): Partial<ClsContext> => {
    const { active }: { active: NamespaceContext | null } = cls || {};

    // eslint-disable-next-line @typescript-eslint/naming-convention
    const { id, _ns_name, ...clsContext } = active || {};
    return clsContext;
  };
}

export function createLogger(
  opts: CreateLoggerOptions = {},
  stream?: pino.DestinationStream,
  // destination?: pino.DestinationStream,
): Logger {
  const cls = createNamespace(`logger/${counter}`);
  counter += 1;

  const {
    platform,
    traceCaller = isDevelopment,
    mixins = [],
    ...userPinoOpts
  } = opts;

  const platformLoggerOptions = getPlatformLoggerOptions(platform);

  const mixin = composeMixins([
    ...mixins,
    createContextMixin(cls),
    traceCaller && callerMixin,
  ]);

  const pinoInstance = stream
    ? pino(
        {
          ...defaultLoggerOptions,
          ...platformLoggerOptions,
          ...userPinoOpts,
          mixin,
        },
        stream,
      )
    : pino({
        ...defaultLoggerOptions,
        ...platformLoggerOptions,
        ...userPinoOpts,
        mixin,
      });

  return Object.create(pinoInstance, {
    cls: {
      value: cls,
      configurable: false,
    },
  });
}

export function expressLogger(
  namespace: Namespace,
  contextId: ClsContext['_contextId'],
  context?: ClsContext['_context'],
): RequestHandler {
  return (req, res, next): void => {
    res.set('x-context-id', contextId);

    namespace.run(() => {
      namespace.set('_contextId', contextId);
      if (context) {
        namespace.set('_context', context);
      }
      next();
    });
  };
}

export function lambdaLogger<T extends (...args: any[]) => any>(
  namespace: Namespace,
  contextId: ClsContext['_contextId'],
  context?: ClsContext['_context'],
): (next: T) => (fn: T) => ReturnType<T> {
  return (fn) => {
    return namespace.runAndReturn(() => {
      namespace.set('_contextId', contextId);
      if (context) {
        namespace.set('_context', context);
      }
      return fn();
    });
  };
}

export function createHttpLogger(
  logger: pino.Logger,
  opts: Omit<pinoHttp.Options, 'logger' | 'customLogLevel'> = {},
): pinoHttp.HttpLogger {
  return pinoHttp({
    ...opts,
    genReqId: () => ({}),
    logger,
    customSuccessMessage(res: ServerResponse) {
      return `${res.statusCode} ${res.statusMessage}`;
    },
    customErrorMessage(err: Error, res: ServerResponse) {
      return `${res.statusCode} ${res.statusMessage}`;
    },
    customLogLevel(res: ServerResponse, err: Error) {
      if (res.statusCode >= 500 || err) {
        return 'error';
      }
      if (res.statusCode >= 400 && res.statusCode < 500) {
        return 'warn';
      }
      return 'trace';
    },
    reqCustomProps(req: IncomingMessage) {
      return { url: req.url };
    },
  });
}
