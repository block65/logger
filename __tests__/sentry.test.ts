import {
  afterEach,
  beforeEach,
  describe,
  expect,
  jest,
  test,
} from '@jest/globals';
import * as sentryModule from '@sentry/node';

const captureException = jest.fn((/* exception, captureContext */) => {
  return 'yes';
});

const addBreadcrumb = jest.fn((/* exception, captureContext */) => {
  return 'yes';
});

const captureMessage = jest.fn((/* message, captureContext */) => {
  return 'yes';
});

const flush = jest.fn((/* message, captureContext */) => {
  return 'yes';
});

jest.unstable_mockModule('@sentry/node', () => {
  return {
    ...sentryModule,
    flush,
    captureException,
    addBreadcrumb,
    captureMessage,
  };
});

describe('Sentry processor', () => {
  beforeEach(() => {
    // jest.resetModules();
    jest.resetAllMocks();
    jest.useFakeTimers('modern');
    jest.setSystemTime(new Date('2009-02-13T23:31:30.000Z'));
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  test('listener one shot', async () => {
    const { Level } = await import('../lib/logger.js');

    const { createSentryListener } = await import('../lib/listener/sentry.js');

    const listener = createSentryListener({
      context: {
        user: {
          id: '1234',
        },
        tags: {
          admin: true,
        },
      },
    });

    listener({
      level: Level.Info,
      time: new Date(),
      msg: 'This happened before the error',
    });

    listener({
      level: Level.Error,
      time: new Date(),
      err: new Error('Oh its a fake error'),
    });

    expect(captureException).toBeCalledTimes(1);
    expect(addBreadcrumb).toBeCalledTimes(1);

    expect(captureException.mock.calls).toMatchSnapshot();
    expect(addBreadcrumb.mock.calls).toMatchSnapshot();
  });

  test('listener as processor', async () => {
    const { createLoggerWithWaitableMock } = await import('./helpers.js');
    const { createSentryListener } = await import('../lib/listener/sentry.js');

    const listener = createSentryListener({
      context: {
        user: {
          id: '1234',
        },
        tags: {
          admin: true,
        },
      },
    });

    const [logger, callback] = createLoggerWithWaitableMock({
      processors: [listener],
    });

    logger.error(new Error('Oops kaboom'));

    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();

    expect(captureException).toBeCalledTimes(1);

    expect(flush).toBeCalledTimes(0);
    expect(captureException.mock.calls).toMatchSnapshot();
    expect(addBreadcrumb.mock.calls).toMatchSnapshot();
  });

  test('listener attach', async () => {
    const { attachSentryListener } = await import('../lib/listener/sentry.js');
    const { createLoggerWithWaitableMock } = await import('./helpers.js');

    const [logger, getLogs] = createLoggerWithWaitableMock();
    attachSentryListener(logger);

    logger.error(new Error('fake'));
    logger.error(new Error('fake'));
    logger.error(new Error('fake'));

    await logger.flush();

    expect(getLogs.mock.calls).toHaveLength(3);

    logger.error(new Error('fake'));
    logger.error(new Error('fake'));
    logger.error(new Error('fake'));

    await logger.end();

    expect(getLogs.mock.calls).toHaveLength(6);
    expect(getLogs.mock.calls).toMatchSnapshot();
  });
});
