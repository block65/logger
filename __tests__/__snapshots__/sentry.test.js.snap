// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sentry processor listener as processor 1`] = `
Array [
  Object {
    "data": Object {
      "err": Object {
        "message": "Oops kaboom",
        "name": "Error",
        "stack": "Error: Oops kaboom
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/sentry.test.js:78:22)",
      },
    },
    "err": [Error: Oops kaboom],
    "level": 50,
    "msg": "Oops kaboom",
    "time": 2009-02-13T23:31:30.000Z,
  },
]
`;

exports[`Sentry processor listener as processor 2`] = `
Array [
  Array [
    [Error: Oops kaboom],
    Object {
      "level": "error",
      "tags": Object {
        "admin": true,
      },
      "user": Object {
        "id": "1234",
      },
    },
  ],
]
`;

exports[`Sentry processor listener as processor 3`] = `Array []`;

exports[`Sentry processor listener attach 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "fake",
          "name": "Error",
          "stack": "Error: fake
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/sentry.test.js:90:22)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
        },
      },
      "err": [Error: fake],
      "level": 50,
      "msg": "fake",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "fake",
          "name": "Error",
          "stack": "Error: fake
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/sentry.test.js:91:22)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
        },
      },
      "err": [Error: fake],
      "level": 50,
      "msg": "fake",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "fake",
          "name": "Error",
          "stack": "Error: fake
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/sentry.test.js:92:22)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
        },
      },
      "err": [Error: fake],
      "level": 50,
      "msg": "fake",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "fake",
          "name": "Error",
          "stack": "Error: fake
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/sentry.test.js:95:22)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
        },
      },
      "err": [Error: fake],
      "level": 50,
      "msg": "fake",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "fake",
          "name": "Error",
          "stack": "Error: fake
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/sentry.test.js:96:22)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
        },
      },
      "err": [Error: fake],
      "level": 50,
      "msg": "fake",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
  Array [
    Object {
      "data": Object {
        "err": Object {
          "message": "fake",
          "name": "Error",
          "stack": "Error: fake
    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/sentry.test.js:97:22)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
        },
      },
      "err": [Error: fake],
      "level": 50,
      "msg": "fake",
      "time": 2009-02-13T23:31:30.000Z,
    },
  ],
]
`;

exports[`Sentry processor listener one shot 1`] = `
Array [
  Array [
    [Error: Oh its a fake error],
    Object {
      "level": "error",
      "tags": Object {
        "admin": true,
      },
      "user": Object {
        "id": "1234",
      },
    },
  ],
]
`;

exports[`Sentry processor listener one shot 2`] = `
Array [
  Array [
    Object {
      "data": undefined,
      "level": "info",
      "message": "This happened before the error",
      "timestamp": 1234567890000,
    },
  ],
]
`;
